contracts/ConsensusUtils.sol

```diff
@@ -287,10 +297,77 @@ contract ConsensusUtils is EternalStorage, ValidatorSet {
+    }
   }
 
   function _delegatedAmountSub(address _address, address _validator, uint256 _amount) internal {
     uintStorage[keccak256(abi.encodePacked("delegatedAmount", _address, _validator))] = uintStorage[keccak256(abi.encodePacked("delegatedAmount", _address, _validator))].sub(_amount);
+    if (uintStorage[keccak256(abi.encodePacked("delegatedAmount", _address, _validator))] == 0) {
+      _delegatorsRemove(_address, _validator);
+    }
+  }
+
+  function delegators(address _validator) public view returns(address[]) {
+    return addressArrayStorage[keccak256(abi.encodePacked("delegators", _validator))];
+  }
+
+  function delegatorsLength(address _validator) public view returns(uint256) {
+    return addressArrayStorage[keccak256(abi.encodePacked("delegators", _validator))].length;
+  }
+
+  function delegatorsAtPosition(address _validator, uint256 _p) public view returns(address) {
+    return addressArrayStorage[keccak256(abi.encodePacked("delegators", _validator))][_p];
+  }
+
+  function isDelegator(address _validator, address _address) public view returns(bool) {
+    for (uint256 i; i < delegatorsLength(_validator); i++) {
+      if (_address == delegatorsAtPosition(_validator, i)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function _setDelegatorsAtPosition(address _validator, uint256 _p, address _address) internal {
+    addressArrayStorage[keccak256(abi.encodePacked("delegators", _validator))][_p] = _address;
+  }
+
+  function _delegatorsAdd(address _address, address _validator) internal {
+    addressArrayStorage[keccak256(abi.encodePacked("delegators", _validator))].push(_address);
+  }
+
+  function _delegatorsRemove(address _address, address _validator) internal {
+    bool found = false;
+    uint256 removeIndex;
+    for (uint256 i; i < delegatorsLength(_validator); i++) {
+      if (_address == delegatorsAtPosition(_validator, i)) {
+        removeIndex = i;
+        found = true;
+      }
+    }
+    if (found) {
+      uint256 lastIndex = delegatorsLength(_validator) - 1;
+      address lastDelegator = delegatorsAtPosition(_validator, lastIndex);
+      if (lastDelegator != address(0)) {
+        _setDelegatorsAtPosition(_validator, removeIndex, lastDelegator);
+      }
+      delete addressArrayStorage[keccak256(abi.encodePacked("delegators", _validator))][lastIndex];
+      addressArrayStorage[keccak256(abi.encodePacked("delegators", _validator))].length--;
+    }
+  }
+
+  function getDelegatorsForRewardDistribution(address _validator, uint256 _rewardAmount) public view returns(address[], uint256[]) {
+    address[] memory _delegators = delegators(_validator);
+    uint256[] memory _rewards = new uint256[](_delegators.length);
+
+    for (uint256 i; i < _delegators.length; i++) {
+      uint256 _amount = delegatedAmount(delegatorsAtPosition(_validator, i), _validator);
+      _rewards[i] = _rewardAmount.mul(_amount).div(getMinStake()).mul(100 - validatorFee(_validator)).div(100);
```

Comment:
here 100 should be the divisibility, lets say 10000 (two digits after the decimal point of a percentage)